src/MemoryTracing/MemoryInfoProducer.cpp

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "MemoryTracing/MemoryInfoProducer.h"
+
+#include <chrono>
+#include <thread>
+
+#include "MemoryTracingUtils.h"
+#include "OrbitBase/Logging.h"
+#include "OrbitBase/ThreadUtils.h"
+#include "capture.pb.h"
+
+namespace orbit_memory_tracing {
+
+using orbit_grpc_protos::SystemMemoryUsage;
+
+void MemoryInfoProducer::Start() {
+  exit_requested_ = false;
+  thread_ = new std::thread(&MemoryInfoProducer::Run, this);
+}
+
+void MemoryInfoProducer::Stop() {
+  exit_requested_ = true;
+  if (thread_ != nullptr && thread_->joinable()) {
+    thread_->join();
+  }
+  delete thread_;
+  thread_ = nullptr;
+}
+
+void MemoryInfoProducer::Run() {
+  orbit_base::SetCurrentThreadName("MIP::Run");
+
+  CHECK(listener_ != nullptr);
+  ProduceSystemMemoryUsageAndSendToListener();
+}
+
+void MemoryInfoProducer::ProduceSystemMemoryUsageAndSendToListener() {
+  std::optional<SystemMemoryUsage> system_memory_usage;
+  std::chrono::time_point scheduled_time = std::chrono::steady_clock::now();
+  while (!exit_requested_) {
+    system_memory_usage = GetSystemMemoryUsage();
+    if (system_memory_usage.has_value()) {
+      listener_->OnSystemMemoryUsage(system_memory_usage.value());
+    }
+    scheduled_time += std::chrono::nanoseconds(sampling_period_ns_);
+    std::this_thread::sleep_until(scheduled_time);
```

Comment:
if sampling_period_ns_ is too big, stop would block for quite considerable time. You can avoid it by protecting stop_requested by a absl::Mutex, and wait on it with timeout with early interruption if flag was set.

Example: https://github.com/google/orbit/blob/main/src/OrbitBase/ThreadPool.cpp#L142

In that case if flag is set - the thread will wake - check the flag and exit.

Reply:
Thanks so much for raising this issue! I changed to use `absl::Mutex` and `AwaitWithDeadline` in the `MemoryInfoProducer`.